File: server/tools/check_spending.js
// The conversational read path into cost tracking (root CLAUDE.md's
// Operational Awareness item 2) — "how much have I spent this month",
// "which model do I use most", "what's left on ElevenLabs". Deliberately no
// dedicated screen for this version (owner's own explicit choice) — this
// tool IS the interface.
//
// core:true (no reliable search-intent text — "how much have I spent"
// doesn't map to a capability search the way "what time is it" maps to
// "time") + meta:true (a live-conversation-only question; a scheduled task
// or a background Job's own turn has no reason to ask this about itself).
// No confirm gate — purely read-only, including the optional balance
// refresh (a GET request against a provider's own account endpoint, same
// risk class as check_myself's own reads).

import * as report from '../cost/report.js';
import { refreshAllBalances } from '../cost/balances.js';

export default {
  name: 'check_spending',
  core: true,
  meta: true,
  description:
    'Check real, measured spend and usage on the paid services you depend on — model API usage, ElevenLabs, Deepgram, or anything ' +
    'else that costs money to run. Answers "how much have I spent," "which model do I use most," "what\'s my ElevenLabs balance." ' +
    'Every number is labelled by kind: MEASURED (what was actually counted), PROVIDER-REPORTED (a real balance figure straight from ' +
    'the provider), or CALCULATED (measured usage × a known price) — never blend these together, and never state a dollar figure ' +
    'for anything this tool reports as having no known price.',
  parameters: {
    type: 'object',
    properties: {
      period: {
        type: 'string',
        enum: ['today', 'month', 'all'],
        description: 'Which window to report on. Defaults to "month" (the natural "how much have I spent this month" scope).',
      },
      refreshBalances: {
        type: 'boolean',
        description:
          'Set true to poll providers for a fresh real balance reading before answering (a real network call to each provider\'s ' +
          'own account endpoint) — otherwise the last known reading is used, which may be stale or entirely absent if none has ' +
          'ever run.',
      },
    },
    required: [],
  },
  async run(args = {}) {
    const period = ['today', 'month', 'all'].includes(args.period) ? args.period : 'month';

    if (args.refreshBalances) {
      try {
        await refreshAllBalances();
      } catch (err) {
        console.error('[check_spending] balance refresh failed — falling back to the last known reading:', err);
      }
    }

    const breakdown =
      period === 'today' ? report.today() : period === 'all' ? report.usageBreakdown('1970-01-01T00:00:00.000Z') : report.monthToDate();
    const mostUsed = report.mostUsed(breakdown.since);
    const balances = report.providerBalances();

    return {
      ok: true,
      period,
      measured: breakdown.measured,
      calculated: breakdown.calculated,
      pricelessGroups: breakdown.pricelessGroups,
      mostUsed,
      providerBalances: balances,
      note:
        balances.length === 0
          ? 'No provider balance has ever been read yet — pass refreshBalances:true to poll now, or wait for the next automatic refresh (every 6 hours).'
          : undefined,
    };
  },
};

File: backend/jarvis/tools/check_spending.py
"""What has actually been spent — the conversational way in to cost tracking.

Deliberately a tool and not a screen, per the owner's own choice: the data and
tool layer first, and a dashboard only once they have used this and know what
they actually want to see.

Answers in three separate parts, and says plainly which of them is missing.
"How much have I spent" with an unpriced model in the roster has no single
honest number, and this returns the shape that admits that rather than one that
hides it.
"""

from __future__ import annotations

from typing import Any

from ..capabilities import CapabilitySpec, Risk

PERIODS = ("today", "month", "week")


def _run(period: str = "month") -> dict[str, Any]:
    from ..cost import report

    wanted = (period or "month").strip().lower()
    if wanted == "today":
        data = report.today()
    elif wanted == "week":
        data = report.last_days(7)
    else:
        wanted, data = "month", report.month_to_date()

    groups = data["measured"]["groups"]
    if not groups:
        return {"ok": True, "period": wanted, "measured": [], "calculated": None,
                "note": "Nothing has been recorded for this period yet."}

    top = max(groups, key=lambda g: g["calls"])
    priceless = data["pricelessGroups"]
    notes = []
    if data["calculated"] is None:
        notes.append("No price is on record for anything used in this period, so there is "
                     "no money figure at all — only real usage counts.")
    elif priceless:
        names = ", ".join(f"{g['modelId'] or g['provider']}" for g in priceless[:5])
        notes.append(f"The money figure covers only what has a price on record. "
                     f"Usage with no known price ({names}) is counted but not costed — "
                     f"say so rather than presenting the total as complete.")

    return {
        "ok": True,
        "period": wanted,
        "since": data["since"],
        "calls": data["eventCount"],
        "measured": groups,
        "calculated": data["calculated"],
        "providerReported": data["providerReported"],
        "pricelessGroups": priceless,
        "mostUsed": {"modelId": top["modelId"], "provider": top["provider"],
                     "calls": top["calls"]},
        "note": " ".join(notes) or "Every number here was counted, not estimated.",
    }


SPEC = CapabilitySpec(
    id="builtin.check_spending",
    name="check_spending",
    description=("Check real recorded usage and spend across every paid service — tokens, "
                 "characters, seconds — for today, this week, or this month. Use it for any "
                 "question about cost, spending, usage or which model gets used most. Some "
                 "usage has no price on record and therefore no money figure: say that "
                 "plainly rather than presenting the total as if it were complete."),
    input_schema={"type": "object", "properties": {
        "period": {"type": "string", "description": '"today", "week" or "month".'}},
        "required": []},
    risk=Risk.LOW,
    handler=_run,
    timeout_s=15.0,
    tags=frozenset({"core", "meta"}),
)
